source/Conversation.h

```diff
@@ -75,17 +75,63 @@ class Conversation {
 	// the user to select; others just automatically continue to another node.
 	// Nodes may also display images or include conditional branches.
 	bool IsChoice(int node) const;
+	// If the given node is a choice node, check how many choices it offers.
 	int Choices(int node) const;
+	// Check if the given conversation node is a conditional branch.
 	bool IsBranch(int node) const;
+	// Check if the given conversation node performs an action.
 	bool IsAction(int node) const;
-	const ConditionSet &Branch(int node) const;
+	const ConditionSet &Conditions(int node) const;
 	const GameAction &GetAction(int node) const;
-	const std::string &Text(int node, int choice = 0) const;
+	const std::string &Text(int node, int element = 0) const;
 	const Sprite *Scene(int node) const;
-	int NextNode(int node, int choice = 0) const;
+	int NextNode(int node, int element = 0) const;
+	// Returns whether the given node should be displayed.
+	// Returns false if:
+	// - The node (or element) is out of range.
+	// - The node is not a choice node, and element is non-zero.
+	// - The node (or element) has conditions and those conditions are not met.
+	// and true otherwise.
+	bool ShouldShowText(const std::map<std::string, int64_t> &vars, int node, int element = 0) const;
+	// Returns true if the given node index is in the range of valid nodes for
+	// this Conversation.
+	// Note: This function only considers actual Conversation nodes to be valid
+	// choices. The negative "outcome" values (ACCEPT, DEFER, etc.) are special
+	// sentinel values, rather than indices of Conversation nodes, and are
+	// therefore considered invalid by this function.
+	bool NodeIsValid(int node) const;
+	// Returns true if the given node index is in the range of valid nodes for
+	// this Conversation *and* the given element index is in the range of valid
+	// elements for the given node.
+	bool ElementIsValid(int node, int element) const;
 
 
 private:
+	// This serves multiple purposes:
+	// - In a regular text node, there's exactly one of these. It contains the
+	//   text data, the index of the next node to unconditionally visit, and,
+	//   optionally, a condition set which, if not met, prevents the text from
+	//   being displayed (without affecting which node is processed next).
+	// - In a choice node, there's one of these for each possible choice,
+	//   containing the text to display, the node the choice leads to, and,
+	//   optionally, the conditions under which to offer the choice.
+	// - In a branch node, there's two of these. The first one contains the
+	//   condition for the branch. If the condition is met, the "next" member
+	//   of the first element is followed. If it's not met, it's the second
+	//   element whose "next" member is followed.
+	class Element {
```

Comment:
I don't think it is unreasonable for this class to be nested (it is done elsewhere in the game).

I'm not sure why it shouldn't be private? This is the only class that needs it in this implementation.

Reply:
We always make classes public I think?

Also while it may not be used yet, there is still the possibility that it may be needed somewhere else later on.

Reply:
It used to be a `std::pair<std::string, int>`. I made it a class so that it would have informative field names (instead of `first` and `second`), and so I could add a third field. It's still basically just a tuple that exists as an implementation detail of `Conversation`.

Both versions of this file have `class Node` nested inside `Conversation`.

Since `Conversation::Element` and `Conversation::Node` both contain references to specific node indices within a specific `Conversation`, it wouldn't really make sense to have either independent of the context of a containing `Conversation`.

Reply:
Anyways, class definitions to the top. You can have it private, but in the top

Reply:
This is placed next to the inner `Node` class, so moving it to the top doesn't make sense in this file. (Unless you also want to move `Node` to the top.)

Reply:
I moved both to the top in my PR to this PR.